solana/solana-ibc/programs/solana-ibc/src/lib.rs

```diff
@@ -191,6 +196,70 @@ pub mod solana_ibc {
         chain.set_stake((validator).into(), amount)
     }
 
+    /// Sets up new fee collector proposal which wont be changed until the new fee collector
+    /// calls `accept_fee_collector_change`. If the method is called for the first time, the fee
+    /// collector would just be set without needing for any approval.
+    ///
+    pub fn setup_fee_collector<'a, 'info>(
+        ctx: Context<'a, 'a, 'a, 'info, SetupFeeCollector<'info>>,
+        new_fee_collector: Pubkey,
+    ) -> Result<()> {
+        let private_storage = &mut ctx.accounts.storage;
+
+        let signer = ctx.accounts.fee_collector.key();
+
+        if private_storage.fee_collector == Pubkey::default() {
+            private_storage.fee_collector = new_fee_collector;
+        } else if signer == private_storage.fee_collector {
+            private_storage.new_fee_collector_proposal = Some(new_fee_collector)
+        } else {
+            return Err(error!(error::Error::InvalidFeeCollector));
+        }
+
+        Ok(())
+    }
+
+    pub fn accept_fee_collector_change<'a, 'info>(
+        ctx: Context<'a, 'a, 'a, 'info, SetupFeeCollector<'info>>,
+    ) -> Result<()> {
+        let private_storage = &mut ctx.accounts.storage;
+
+        let signer = ctx.accounts.fee_collector.key();
+
+        if let Some(new_admin) = private_storage.new_fee_collector_proposal {
+            if signer != new_admin {
+                return Err(error::Error::InvalidFeeCollector)
+                    .map_err(|err| error!((&err)));
+            }
+            private_storage.fee_collector = new_admin;
+            private_storage.new_fee_collector_proposal = None;
+        } else {
+            return Err(error!(error::Error::FeeCollectorChangeProposalNotSet));
+        }
+
+        Ok(())
+    }
+
+    pub fn collect_fees<'a, 'info>(
+        ctx: Context<'a, 'a, 'a, 'info, CollectFees<'info>>,
+    ) -> Result<()> {
+        let fee_collector_balance = ctx.accounts.fee_account.lamports();
+
+        if fee_collector_balance < MINIMUM_FEE_TO_COLLECT {
+            return Err(error!(error::Error::InsufficientFeesToCollect));
+        }
+
+        let total_fees_collected =
+            fee_collector_balance - MINIMUM_FEE_TO_COLLECT;
```

Comment:
Why don’t we grab everything?

Reply:
we keep 1 sol as minimum balance for rent exempt and enough balance for refunds.